cpp/src/Glacier2Lib/NullPermissionsVerifier.cpp

```diff
@@ -26,114 +26,63 @@ class NullSSLPermissionsVerifier : public Glacier2::SSLPermissionsVerifier
 {
 public:
 
-    virtual bool
-    authorize(Glacier2::SSLInfo, string&, const Ice::Current&) const
+    bool authorize(Glacier2::SSLInfo, string&, const Ice::Current&) const final
     {
         return true;
     }
 };
 
-class Init
-{
-public:
-
-    Init(const CommunicatorPtr&, const string&, const vector<string>&);
-
-private:
-
-    string checkPermissionVerifier(const string&);
-    void createObjects();
-
-    const CommunicatorPtr _communicator;
-    ObjectAdapterPtr _adapter;
+}
 
-    Identity _nullPVId;
-    Identity _nullSSLPVId;
-};
+namespace Glacier2Internal
+{
 
-Init::Init(const CommunicatorPtr& communicator, const string& category, const vector<string>& props) :
-    _communicator(communicator)
+void
+setupNullPermissionsVerifier(
+    const CommunicatorPtr& communicator,
+    const string& category,
+    const vector<string>& permissionsVerifierPropertyNames)
 {
-    _nullPVId.name = "NullPermissionsVerifier";
-    _nullPVId.category = category;
+    const Ice::Identity nullPermissionsVerifierId {"NullPermissionsVerifier", category};
+    const Ice::Identity nullSSLPermissionsVerifierId {"NullSSLPermissionsVerifier",  category};
+
+    const Ice::PropertiesPtr properties = communicator->getProperties();
 
-    _nullSSLPVId.name = "NullSSLPermissionsVerifier";
-    _nullSSLPVId.category = category;
+    shared_ptr<Glacier2::PermissionsVerifier> nullPermissionsVerifier;
+    shared_ptr<Glacier2::SSLPermissionsVerifier> nullSSLPermissionsVerifier;
 
-    Ice::PropertiesPtr properties = _communicator->getProperties();
-    for(vector<string>::const_iterator p = props.begin(); p != props.end(); ++p)
+    for (const auto& propertyName : permissionsVerifierPropertyNames)
     {
-        string val = properties->getProperty(*p);
-        if(!val.empty())
+        string propertyValue = properties->getProperty(propertyName);
+        if (!propertyValue.empty())
         {
-            //
-            // Check permission verifier proxy. It returns a non-empty
-            // value with the new stringified proxy if the property
-            // needs to be rewritten.
-            //
-            val = checkPermissionVerifier(val);
-            if(!val.empty())
+            ObjectPrx prx(communicator, propertyValue);
+            if (prx->ice_getIdentity() == nullPermissionsVerifierId && nullPermissionsVerifier == nullptr)
+            {
+                nullPermissionsVerifier = make_shared<NullPermissionsVerifier>();
+            }
+            else if (prx->ice_getIdentity() == nullSSLPermissionsVerifierId && nullSSLPermissionsVerifier == nullptr)
             {
-                properties->setProperty(*p, val);
+                nullSSLPermissionsVerifier = make_shared<NullSSLPermissionsVerifier>();
             }
         }
     }
-}
 
-string
-Init::checkPermissionVerifier(const string& val)
-{
-    // Check if it's in proxy format
-    try
+    if (nullPermissionsVerifier != nullptr || nullSSLPermissionsVerifier != nullptr)
```

Comment:
I would avoid these "!= nullptr". It's more idiomatic to just convert these shared_ptr to bool.